src/ILCompiler.ReadyToRun/src/Compiler/DependencyAnalysis/ReadyToRunCodegenNodeFactory.cs

```diff
@@ -12,32 +13,209 @@
+
+        public NodeFactory(CompilerTypeSystemContext context,
+            CompilationModuleGroup compilationModuleGroup,
+            NameMangler nameMangler,
+            MetadataManager metadataManager)
+        {
+            TypeSystemContext = context;
+            CompilationModuleGroup = compilationModuleGroup;
+            Target = context.Target;
+            NameMangler = nameMangler;
+            MetadataManager = metadataManager;
+
+            CreateNodeCaches();
+        }
+
+        public CompilerTypeSystemContext TypeSystemContext { get; }
+
+        public TargetDetails Target { get; }
+
+        public CompilationModuleGroup CompilationModuleGroup { get; }
+
+        public NameMangler NameMangler { get; }
+
+        public MetadataManager MetadataManager { get; }
+
+        public bool MarkingComplete => _markingComplete;
+
+        public void SetMarkingComplete()
+        {
+            _markingComplete = true;
+        }
+
+        private void CreateNodeCaches()
+        {
+            _typeSymbols = new NodeCache<TypeDesc, IEETypeNode>(CreateNecessaryTypeNode);
+            _constructedTypeSymbols = new NodeCache<TypeDesc, IEETypeNode>(CreateConstructedTypeNode);
+            _methodEntrypoints = new NodeCache<MethodDesc, IMethodNode>(CreateMethodEntrypointNode);
+            _genericReadyToRunHelpersFromDict = new NodeCache<ReadyToRunGenericHelperKey, ISymbolNode>(CreateGenericLookupFromDictionaryNode);
+            _genericReadyToRunHelpersFromType = new NodeCache<ReadyToRunGenericHelperKey, ISymbolNode>(CreateGenericLookupFromTypeNode);
+
+            _readOnlyDataBlobs = new NodeCache<ReadOnlyDataBlobKey, BlobNode>(key =>
+            {
+                return new BlobNode(key.Name, ObjectNodeSection.ReadOnlyDataSection, key.Data, key.Alignment);
+            });
+        }
+
+        public abstract void AttachToDependencyGraph(DependencyAnalyzerBase<NodeFactory> graph);
+
+        protected abstract IMethodNode CreateMethodEntrypointNode(MethodDesc method);
+
+        protected abstract IEETypeNode CreateNecessaryTypeNode(TypeDesc type);
+
+        protected abstract IEETypeNode CreateConstructedTypeNode(TypeDesc type);
+
+        protected abstract ISymbolNode CreateGenericLookupFromDictionaryNode(ReadyToRunGenericHelperKey helperKey);
+
+        protected abstract ISymbolNode CreateGenericLookupFromTypeNode(ReadyToRunGenericHelperKey helperKey);
+
+        private NodeCache<TypeDesc, IEETypeNode> _typeSymbols;
+
+        public IEETypeNode NecessaryTypeSymbol(TypeDesc type)
+        {
+            return _typeSymbols.GetOrAdd(type);
+        }
+
+        private NodeCache<TypeDesc, IEETypeNode> _constructedTypeSymbols;
+
+        public IEETypeNode ConstructedTypeSymbol(TypeDesc type)
+        {
+            return _constructedTypeSymbols.GetOrAdd(type);
+        }
+
+        protected NodeCache<MethodDesc, IMethodNode> _methodEntrypoints;
+
+        public IMethodNode MethodEntrypoint(MethodDesc method)
```

Comment:
I would love to figure out how to get rid of this method that has a very lame implementation in the R2RCgNF. It's used from only one place in CPAOT - to translate MethodDesc's to previously compiled nodes in EnumerateCompiledMethods - and that should be easy to fix, perhaps by explicitly iterating over _localMethods and _importMethods. I'm not saying you should fix all this as part of your change which is complex enough already, I'm just curious about your thoughts on the subject; I or Simon can look at the actual refactoring if we decide it's the right way to go.

Reply:
We also use it for rooting, but yeah, it should be possible to do this. I'll add a `TODO-REFACTOR` for this and do it in another pull request.